### PR TITLE
Implement Datastream override feature + Unit tests + Integration tests

### PR DESCRIPTION
**Pull request by addb**
Implement datastream override feature.

Expected implementation:

```
data = {
    "xdm" : {
        "xdmKey" : "value"
    },
    "data" : {
        "key" : "value",
     },
    "config": {
        "datastreamIdOverride" : "<DATASTREAM_ID_OVERRIDE_VALUE>"
    }
} 

''' sendEvent with datastream Id override
sendEvent(data)
```

```
data = {
    "xdm" : {
        "xdmKey" : "value"
    },
    "data" : {
        "key" : "value",
     },
    "config": {
        "datastreamConfigOverride" : {
            "com_adobe_analytics": {
                "reportSuites": [
                    "reportSuite1",
                    "reportSuite2"
                ]
            },
            "com_adobe_target": {
                "enabled": false
            },
        }
    }
} 

''' sendEvent with datastream config override
sendEvent(data)
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/main/edge/edgeRequestWorker.brs [hunk 1]
```diff
@@ -132,16 +133,42 @@ function _adb_EdgeRequestWorker() as object
             return responseArray
         end function,
 
-        _processRequest: function(eventData as object, ecid as string, configId as string, requestId as string, path as string, edgeDomain = invalid as dynamic) as object
-            requestBody = m._createEdgeRequestBody(eventData, ecid)
+        _processRequest: function(eventData as object, ecid as string, datastreamId as string, requestId as string, path as string, meta = {} as object, edgeDomain = invalid as dynamic) as object
+            if not _adb_isEmptyOrInvalidMap(eventData.config)
+                config = eventData.config
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:138]: Can you extract this logic into a seperate function?

File: code/main/edge/edgeRequestWorker.brs [hunk 2]
```diff
@@ -132,16 +133,42 @@ function _adb_EdgeRequestWorker() as object
             return responseArray
         end function,
 
-        _processRequest: function(eventData as object, ecid as string, configId as string, requestId as string, path as string, edgeDomain = invalid as dynamic) as object
-            requestBody = m._createEdgeRequestBody(eventData, ecid)
+        _processRequest: function(eventData as object, ecid as string, datastreamId as string, requestId as string, path as string, meta = {} as object, edgeDomain = invalid as dynamic) as object
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:136]: I'm not familiar with the function's usage, but I prefer appending new default parameters to the end of the parameter list. Regarding "meta" do we require a default value considering we perform error checks within the function?

File: code/main/edge/edgeRequestWorker.brs [hunk 3]
```diff
@@ -132,16 +133,42 @@ function _adb_EdgeRequestWorker() as object
...
+            if not _adb_isEmptyOrInvalidMap(eventData.config)
+                config = eventData.config
+                sdkConfig = invalid
+                configOverrides = invalid
 
-            url = _adb_buildEdgeRequestURL(configId, requestId, path, edgeDomain)
+                if meta = invalid
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:142]: if _adb_isEmptyOrInvalidMap(meta)
   meta = {}
end if

File: code/main/edge/edgeRequestWorker.brs [hunk 4]
```diff
@@ -132,16 +133,42 @@ function _adb_EdgeRequestWorker() as object
...
+            requestBody = m._createEdgeRequestBody(eventData, ecid, meta)
+
+            url = _adb_buildEdgeRequestURL(datastreamId, requestId, path, edgeDomain)
             _adb_logVerbose("EdgeRequestWorker::_processRequest() - Processing Request with url:(" + chr(10) + FormatJson(url) + chr(10) + ") with payload:(" + chr(10) + FormatJson(requestBody) + chr(10) + ")")
             networkResponse = _adb_serviceProvider().networkService.syncPostRequest(url, requestBody)
             return networkResponse
         end function
 
-        _createEdgeRequestBody: function(eventData as object, ecid as string) as object
+        _createEdgeRequestBody: function(eventData as object, ecid as string, meta = invalid as object) as object
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:171]: Can you remove the default value?

File: code/main/edge/edgeRequestWorker.brs [hunk 5]
```diff
@@ -153,9 +196,20 @@ function _adb_EdgeRequestWorker() as object
...
+            end if
+
             return requestBody
         end function,
 
+        _getSdkConfigPayload: function(originalDatastreamId as string) as object
+            sdkConfig = {}
+            sdkConfig.datastream = {}
+            sdkConfig.datastream.original = originalDatastreamId
+        return sdkConfig
```
Comment (yangyansong-adbe) [inline code/main/edge/edgeRequestWorker.brs:210]: Can we adjust the code format?

File: code/main/edge/edgeRequestWorker.brs [hunk 6]
```diff
@@ -132,16 +133,58 @@ function _adb_EdgeRequestWorker() as object
...
+            requestBody = m._createEdgeRequestBody(eventData, ecid, meta)
+
+            url = _adb_buildEdgeRequestURL(datastreamId, requestId, path, edgeDomain)
             _adb_logVerbose("EdgeRequestWorker::_processRequest() - Processing Request with url:(" + chr(10) + FormatJson(url) + chr(10) + ") with payload:(" + chr(10) + FormatJson(requestBody) + chr(10) + ")")
             networkResponse = _adb_serviceProvider().networkService.syncPostRequest(url, requestBody)
             return networkResponse
         end function
 
-        _createEdgeRequestBody: function(eventData as object, ecid as string) as object
+        _getDatastreamId: function(config as object, originalDatastreamId as String) as string
```
Comment (yangyansong-adbe) [inline code/main/edge/edgeRequestWorker.brs:154]: nit: can we rename it to `_getOverriddenDatastreamId` ?

File: code/main/edge/edgeRequestWorker.brs [hunk 7]
```diff
@@ -132,16 +133,58 @@ function _adb_EdgeRequestWorker() as object
...
+            if _adb_isEmptyOrInvalidMap(config)
+                return originalDatastreamId
+            end if
+
+            if not _adb_isEmptyOrInvalidString(config.datastreamIdOverride)
```
Comment (yangyansong-adbe) [inline code/main/edge/edgeRequestWorker.brs:159]: I'm not 100% sure, but I guess `config.datastreamIdOverride` could cause errors if the key doesn't exist. Can we use `config["datastreamIdOverride"]` or call [Lookup(key as String) as Dynamic](https://developer.roku.com/docs/references/brightscript/interfaces/ifassociativearray.md#lookupkey-as-string-as-dynamic) function instead?

The same concern about the below code: 
```brightscript
            if not _adb_isEmptyOrInvalidMap(config.datastreamConfigOverride)
                meta["configOverrides"] = config.datastreamConfigOverride
            end if

            if not _adb_isEmptyOrInvalidString(config.datastreamIdOverride)
                ''' Genrate sdkConfig payload with original datastreamId
                meta["sdkConfig"] = m._getSdkConfigPayload(originalDatastreamId)
            end if
```

File: code/main/edge/edgeRequestWorker.brs (hunk 6 above)
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:154]: So this method would either send overriddenDatastremId or the original if not present. That's why named it like this.

File: code/main/edge/edgeRequestWorker.brs (hunk 7 above)
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:159]: I have added an test with config and it is working fine. So It shouldn't be a problem.

File: code/main/edge/edgeRequestWorker.brs (hunk 5 above)
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:210]: Are you thinking of something like this:

```
sdkConfig = {"datastream" : { "original": originalDatastreamId } } 
```

File: code/main/edge/edgeRequestWorker.brs [hunk 8]
```diff
@@ -132,16 +133,58 @@ function _adb_EdgeRequestWorker() as object
...
         end function,
 
-        _processRequest: function(eventData as object, ecid as string, configId as string, requestId as string, path as string, edgeDomain = invalid as dynamic) as object
-            requestBody = m._createEdgeRequestBody(eventData, ecid)
+        _processRequest: function(eventData as object, ecid as string, datastreamId as string, requestId as string, path as string, meta as object, edgeDomain as dynamic) as object
+            ''' Append config overrides to meta if set in eventData.config
+            meta = m._appendConfigOverridesToMeta(meta, eventData.config, datastreamId)
 
-            url = _adb_buildEdgeRequestURL(configId, requestId, path, edgeDomain)
+            ''' Override the datastreamId if set in eventData.config
```
Comment (praveek) [inline code/main/edge/edgeRequestWorker.brs:140]: Can you reword this comment to reflect what the function is doing?

Comment (praveek) [reply code/main/edge/edgeRequestWorker.brs:140]: Can you also update this function comment?
https://github.com/adobe/aepsdk-roku/pull/152/files#diff-b7b00e7cee0f49d4f627be8f64df686c4e8410614c93830baac78cf755b18c5bL28

File: code/tests/edge/test_edgeRequestWorker.brs [hunk 9]
```diff
@@ -127,14 +127,419 @@ sub TC_adb_EdgeRequestWorker_processRequest_valid_response()
...
+                    key: "value"
+                }
+            }
+        ]
+        UTF_assertEqual(expectedEventsArray, jsonObj.events, "Expected != actual (Events payload in the request)")
+        return _adb_NetworkResponse(200, "response body")
+    end function
+
+    worker = _adb_EdgeRequestWorker()
+    networkResponse = worker._processRequest({ xdm: { key: "value" }, config: { key: "value"} }, "ecid_test", "config_id", "request_id", "/ee/v1/interact", invalid, invalid)
```
Comment (praveek) [inline code/tests/edge/test_edgeRequestWorker.brs:534]: nit: Can you change config from 
```
config: {
   key: "value" 
} 
```
to 

```
config: {
  key1: "value1" 
}
```

It is little confusing when xdm and config has same key and value.

Comment (praveek) [reply code/tests/edge/test_edgeRequestWorker.brs:534]: Can you also add a test scenario where the data types for datastreamOverrideId and datastreamConfigOverride are not the types that are expected?

**Review by praveek — APPROVED** [review]

**Review by yangyansong-adbe — APPROVED** [review]

**Pull request merged**